source/Test.cpp

```diff
@@ -0,0 +1,215 @@
+/* Test.cpp
+Copyright (c) 2019-2020 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Test.h"
+
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "MainPanel.h"
+#include "Panel.h"
+#include "Planet.h"
+#include "PlanetPanel.h"
+#include "PlayerInfo.h"
+#include "Ship.h"
+#include "System.h"
+#include "TestData.h"
+#include "UI.h"
+
+#include <SDL2/SDL.h>
+
+#include <algorithm>
+#include <map>
+#include <numeric>
+#include <stdexcept>
+
+using namespace std;
+
+
+
+namespace{
+	const map<Test::Status, const string> STATUS_TO_TEXT = {
+		{Test::Status::ACTIVE, "active"},
+		{Test::Status::BROKEN, "broken"},
+		{Test::Status::KNOWN_FAILURE, "known failure"},
+		{Test::Status::MISSING_FEATURE, "missing feature"},
+	};
+	
+	const map<Test::TestStep::Type, const string> STEPTYPE_TO_TEXT = {
+		{Test::TestStep::Type::ASSIGN, "assign"},
+		{Test::TestStep::Type::ASSERT, "assert"},
+		{Test::TestStep::Type::BRANCH, "branch"},
+		{Test::TestStep::Type::INJECT, "inject"},
+		{Test::TestStep::Type::INPUT, "input"},
+		{Test::TestStep::Type::INVALID, "invalid"},
+		{Test::TestStep::Type::LABEL, "label"},
+		{Test::TestStep::Type::NAVIGATE, "navigate"},
+		{Test::TestStep::Type::WATCHDOG, "watchdog"},
+	};
+	
+	template<class K, class... Args>
+	string ExpectedOptions(const map<K, const string, Args...> &m)
+	{
+		if(m.empty())
+			return "no options supported";
+		
```

Comment:
I think this code also needs an exception for dealing with maps that only contain 1 element:
- Functional: I don't know if accumulate can handle the begin-range element being behind the end-range element.
- Textual: here should not be an `or` in the string when there is only 1 element.

Reply:
Handled in d7150a6

Reply:
I didn't bother with those since this is file-local only and not truly "generic" -- we have the guarantee that any mapping passed to it is non-empty and also has more than 2 elements, so the beginning and end iterators are well-formed.
As a file-local method, anyone adding new calls to it is trivially able to view and/or modify the implementation.

I added the empty check mostly for sanity - we dereference `begin` - and because I considered it might result in better codegen since the compiler will know more about the map state after it. (I didn't verify that assumption, though.)